File: #1026.maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {

        int result = 0;
        ancestorDiff(root->left, root->val, root->val, result);
        ancestorDiff(root->right, root->val, root->val, result);
        return result;
    }

private:
    void ancestorDiff(TreeNode* root, int max, int min, int& result) {

        if (!root) {
            return;
        }

        result = std::max(result, std::abs(root->val - max));
        result = std::max(result, std::abs(root->val - min));
        max = std::max(max, root->val);
        min = std::min(min, root->val);
        ancestorDiff(root->left, max, min, result);
        ancestorDiff(root->right, max, min, result);
    }
};

File: #1026.maximum-difference-between-node-and-ancestor.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxAncestorDiff = function(root) {
    let result = 0;
    const dfs = (node, {max, min}) => {
        if (!node) {
            return;
        }

        const { val, left, right } = node;
        result = Math.max(result, Math.abs(max - val), Math.abs(min - val));
        max = Math.max(max, val);
        min = Math.min(min, val);
        dfs(left, {max, min});
        dfs(right, {max, min});
    };

    dfs(root, {max: root.val, min: root.val});
    return result;
};
